pkg/policy/groups/helpers.go

```diff
@@ -91,6 +99,71 @@ func createDerivativeCNP(ctx context.Context, cnp *cilium_v2.CiliumNetworkPolicy
 	return derivativeCNP, nil
 }
 
+// createDerivativeCCNP will return a new CCNP based on the given rule.
+func createDerivativeCCNP(ctx context.Context, cnp *cilium_v2.CiliumNetworkPolicy) (*cilium_v2.CiliumClusterwideNetworkPolicy, error) {
+	ccnp := &cilium_v2.CiliumClusterwideNetworkPolicy{
+		TypeMeta:            cnp.TypeMeta,
+		ObjectMeta:          cnp.ObjectMeta,
+		CiliumNetworkPolicy: cnp,
+		Status:              cnp.Status,
+	}
+
+	// CCNP informer may provide a CCNP object without APIVersion or Kind.
+	// Setting manually to make sure that the derivative policy works ok.
+	derivativeCCNP := &cilium_v2.CiliumClusterwideNetworkPolicy{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      getDerivativeName(ccnp),
+			Namespace: ccnp.ObjectMeta.Namespace,
+			OwnerReferences: []v1.OwnerReference{{
+				APIVersion:         cilium_v2.SchemeGroupVersion.String(),
+				Kind:               cilium_v2.CCNPKindDefinition,
+				Name:               ccnp.ObjectMeta.Name,
+				UID:                ccnp.ObjectMeta.UID,
+				BlockOwnerDeletion: &blockOwnerDeletionPtr,
+			}},
+			Labels: map[string]string{
+				parentCNP:  string(ccnp.ObjectMeta.UID),
+				cnpKindKey: cnpKindName,
+			},
+		},
+		CiliumNetworkPolicy: &cilium_v2.CiliumNetworkPolicy{},
+	}
+
+	var (
+		rules api.Rules
+		err   error
+	)
+
+	rules, err = ccnp.Parse()
+
+	if err != nil {
+		// We return a valid pointer for derivative policy here instead of nil.
+		// This object is used to get generated name for the derivative policy
+		// when updating the status of the network policy.
+		return derivativeCCNP, fmt.Errorf("cannot parse CCNP: %v", err)
+	}
+
+	derivativeCCNP.Specs = make(api.Rules, len(rules))
+	for i, rule := range rules {
+		if rule.RequiresDerivative() {
+			derivativeCCNP.Specs[i] = denyEgressRule()
+		}
+	}
+
+	for i, rule := range rules {
+		if !rule.RequiresDerivative() {
+			derivativeCCNP.Specs[i] = rule
+			continue
+		}
+		newRule, err := rule.CreateDerivative(ctx)
+		if err != nil {
+			return derivativeCCNP, err
+		}
+		derivativeCCNP.Specs[i] = newRule
+	}
```

Comment:
This can be done as a follow up but this section is the same as done in the `createDerivativeCNP`. This can be done as part of a dedicated function.

I can see that `createDerivativeCCNP` is returning a CCNP on line 160 (in case of a error) which would make it difficult to have a dedicated function for this section. That's why I'm saying that it can be done as a follow up because it will require changes in the caller of the `createDerivativeCCNP` to ignore the returned CCNP in case of an error. We shouldn't use the returned values if `err != nil`.

Reply:
Make sense. I will create a follow-up issue once this is merged to address all the remaining issues.